### ISSUE TITLE
Watching existing child process PIDs

### ISSUE THREAD

**Issue by leonerd** [libuv/libuv#3100]
I'm unsure if this is a documentation issue or a feature request.

I have existing code that does the `fork` (and maybe `exec` or maybe not) parts of creating a child process, and I wish to use `libuv` to watch those processes for exit. Can I do this? I couldn't see anything in docs - there doesn't appear to be a "uv_child_t" type that I can give an existing PID to, and all of the ways I can see to create a "uv_process_t" start off by my supplying a new command to be exec'ed, and it does the whole dance for me.

I currently have a very hacky workaround, which involves my watching for `SIGCHLD` directly, and on receipt of such a signal I try to `waitpid()` specifically on every child process PID I still have outstanding. While that works in small cases, it doesn't feel like it will scale very well. Anything else would interfere with libuv's own child process reaping.

If this ability does exist in libuv, then I wasn't able to find it in the docs, so perhaps more attention needs to be drawn to it.

Alternatively if not, can this be considered a feature request to add such? Maybe either an alternative to "uv_spawn()" which still uses a uv_process_t, or else a whole new handle type (uv_child_t?) that just wraps a preëxisting process created elsewhere, that libuv will just watch for exit.

**Comment by shadowcat-mst** [comment]
I've also run into situations where I get a pid back from other code, often not code that I control, but would like to integrate detecting the process' completion into event loop usage, so either pointing out which feature we've missed or adding it to the backlog would be much appreciated.

**Comment by bnoordhuis** [comment]
> I currently have a very hacky workaround, which involves my watching for SIGCHLD directly, and on receipt of such a signal I try to waitpid() specifically on every child process PID I still have outstanding. While that works in small cases, it doesn't feel like it will scale very well.

That is exactly what libuv does: 5c00a0ee331ed77ee91a75ab33612afeb1a91088 :-)

Libuv doesn't currently support what you're asking and I'm undecided if it makes sense for libuv to add it. How would it work on Windows? `uv_spawn()` on Windows is a rather complicated affair.

One issue that comes to mind: it makes the race condition that 041d60e1416335a817f735c41e033d27bd8f7ade fixed possible again because there's no way for libuv to know for sure if the child process has execve'd or not, not unless you copy all the safeguards that libuv has in place.

**Comment by leonerd** [comment]
> That is exactly what libuv does: [5c00a0e](https://github.com/libuv/libuv/commit/5c00a0ee331ed77ee91a75ab33612afeb1a91088) :-)

Yes; I found that code and that inspired my currently-deployed workaround. I just felt that if libuv is going to implement the behaviour of keeping a list of all interesting PIDs and linearly scan them, it might as well allow me to add my own to that list rather than making me duplicate the work.

> Libuv doesn't currently support what you're asking and I'm undecided if it makes sense for libuv to add it. How would it work on Windows? `uv_spawn()` on Windows is a rather complicated affair.

libuv already offers quite a number of things that just don't work on Windows - e.g. most of the subtleties of signals.

Glib, for example, solves this by offering a bunch of "glib_unix_..." functions; saying "well here's some things that are easy to do on UNIX but you promise not to complain that they don't work on Windows". If I am `fork()`ing myself, I've already taken that pill already.

> One issue that comes to mind: it makes the race condition that [041d60e](https://github.com/libuv/libuv/commit/041d60e1416335a817f735c41e033d27bd8f7ade) fixed possible again because there's no way for libuv to know for sure if the child process has execve'd or not, not unless you copy all the safeguards that libuv has in place.

That's basically about early-after-fork races. If we're integrating with some external library or subsystem that just gives us PIDs, it's up to that library to do the right thing there, and we just have to hope that it does. I don't see how adding a hypothetical `uv_unix_watch_pid(...)` affects that.

**Comment by bnoordhuis** [comment]
A guiding principle in libuv's design is "must be reasonably supported on all platforms, or a reasonable facsimile must be provided on platforms where it isn't." Signals are in the second category.

You're welcome to open a PR (because expectation management: I don't see any maintainers working on this anytime soon) but keep in mind you can't just ignore Windows.

**Comment by skissane** [comment]
The same issue can occur on Windows. Windows has APIs to create a process in a different security context (CreateProcessWithTokenW, CreateProcessAsUser, CreateProcessWithLogonW). Right now, libuv has no support for calling these. If one wants to use them with libuv, one has three basic options:

1. Enhance libuv to add support for calling these APIs
2. Enhance libuv to allow importing an existing child process (in Windows' case via a handle not a PID) and thereafter having libuv manage it – essentially what @leonerd is asking for in this issue
3. Managing those child processes manually – essentially what @leonerd is doing now

The reasons for wanting to do it may be a bit different between Unix-likes and Windows, but the want to have libuv manage externally created child processes exists on both platforms.

**Comment by leonerd** [comment]
How do we progress this issue? I'm afraid I couldn't help with the Windows-shaped API but I could chat with some folks on that and see if we can find a unified API shape that could work.
